### DISCUSSION TITLE
Strategies for complex pydantic models and their unions

### DISCUSSION THREAD

**Discussion by jajcayn** [pydantic/pydantic#8822] [Question]
Hey folks, I am working on a project with FastAPI + Pydantic, the whole app is meant to work with patient data. I need the strictest validation possible since once the data are in the DB, I absolutely have to know they are correct. The issue I have can be boiled down to a simple example:
- Imagine a DB table `medical data`. Since I aim to support any medical data, I cannot create a column for each data type, so instead, I have a DB table like this: "data type": str (e.g., Systolic pressure), "data value": float, "data unit": string and others, but they do not matter.
- Now, for data types that I directly ssupport, I need a float range validation (e.g. pressure must be between 1-250)
- I also need a default schema for data types that do not need or, I do not require validation
- Since this is classic CRUD, I need a schema for response (more data), a schema for POST, and a schema for update (all optional)
- Now for the tricky part - I need Unions of those models and some reasonable extensibility to the future (adding more models on the go)

A minimal example could be something like this:
```python

class BaseMedicalData(BaseModel):
    data_name: str
    data_value: float
    data_unit: str
    ....

class MedicalDataResponse(BaseMedicalData):
    uuid: UUID4
    ....

class CreateMedicalData(BaseMedicalData):
    pass

class UpdateMedicalData(BaseModel):
    data_name: str | None = None
    data_value: float | None = None
    data_unit: str | None = None
    ....

class BaseSystolicPressure(BaseMedicalData):
    data_name: Literal["systolic pressure"]
    data_value: Annotated[float, Field(ge=1, le=250)]
    data_unit: Literal["mmHg"]

...
```

And so on, you get the idea. The problem is, that I have at least 50 of these specific measurements, and need to do a strict range validation. If I needed define 3 models per data type, I would have 150 pydantic models, plus when I am returning a list of all data, the Union of these models would be hell to write by hand. 

So what I have so far is:
- define only three "default models" (like my example)
- for each type define just a `BaseModel` with a specific validation range and literal name and unit
- create manually a dictionary of "custom models": `custom_models: {"systolic pressure": SystolicPressureSchema, ...}` and then use my "infamous function" using `create_model`:

```python
def create_models_dynamically(specific_schemas, base_create_schema, base_response_schema, base_update_schema):
    get_models = {}
    post_models = {}
    put_models = {}
    for key, base_schema in specific_schemas.items():
        model_name = key.title().replace(" ", "").replace("-", "")
        get_models[key] = create_model(f"{model_name}Response", __base__=(base_schema, base_response_schema))
        post_models[key] = create_model(f"Create{model_name}", __base__=(base_schema, base_create_schema))
        put_models[key] = create_model(f"Update{model_name}", <# iterate over fields and wrap in Optional>, __base__=(base, schema, base_update_schema))
    return get_models, post_models, put_models
```
This works nicely because I have a dictionary with all my models (150) with a correct validation setup. However, then when it comes to using the Union of these models in FastAPI endpoints, the only thing I could come up with is:

```python
UnionGetModels = Union[tuple(get_models.values()) + tuple([DefaultResponseSchema])]
```

Another problem, when I want to validate against particular model inside FastAPI endpoint I need to do something like:
```
data = <incoming data via endpoint>
validated = get_models.get(data.data_name, DefaultResponseSchema).model_validate(data)
```

The Final problem is when there is actually a validation error, the error log is HUGE, because it prints all possible validation errors.
Now, I am familiar with discriminated unions, but given the way how I create Unions dynamically I am not sure whether this would work for my case. Plus, I am not sure how to work with my defaults because AFAIK discriminated union needs the discriminator to be of literal type.
Also, a very minor issue, but still, since the Union created like this is not registered by pylance (I use VSCode), I do not get type hints, nor auto-complete hints on any object that is typed as this Union.

So my question is: how to do this elegantly, explicitely, and with all the perks of pydantic and python typing? Or is my use case so bizarre there is no way how to do it, and my hacky way is the best there is currently?

Cheers!

**Comment by GDYendell** [comment]
You could insert the `Literal` in `create_model` and use that as the discriminator when creating the `Union`. Something like this maybe (replacing your bases with literal fields for a simpler demo)

```python
models: dict[str, BaseModel] = {
    name: create_model(
        f"{name}Response",
        __config__=ConfigDict(extra="forbid"),
        uuid=(int, 0),
        type=(Literal[f"{name}Response"], "{name}Response"),
        **{name: (str, "")}  # Add field "X", "Y" or "Z" for each model
    )
    for name in ["X", "Y", "Z"]
}

ResponseUnion = Annotated[Union[tuple(models.values())], Field(discriminator="type")]

pprint(TypeAdapter(ResponseUnion).json_schema())

models["X"].model_validate({"type": "XResponse", "uuid": 1, "X": "x"})  # Works
models["X"].model_validate({"type": "XResponse", "uuid": 1, "Y": "y"})  # Fails
```

Output:
```
Traceback (most recent call last):
  File "/workspaces/pvi2/_pydantic.py", line 100, in <module>
    models["X"].model_validate({"type": "XResponse", "uuid": 1, "Y": "y"})
  File "/venv/lib/python3.10/site-packages/pydantic/main.py", line 503, in model_validate
    return cls.__pydantic_validator__.validate_python(
pydantic_core._pydantic_core.ValidationError: 1 validation error for XResponse
Y
  Extra inputs are not permitted [type=extra_forbidden, input_value='y', input_type=str]
    For further information visit https://errors.pydantic.dev/2.5/v/extra_forbidden
```

This might improve your error messages.

You could find it more useful to use a [callable](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-callable-discriminator) discriminator with `Discriminator` to create tagged unions if you need to be more flexible than a `Literal` field on every class. You do still have to create a unique `Tag` string for each union member where you construct the Union, but you can perform more complex logic to tell pydantic how to discriminate between types. I think this also makes it easier (or possible) to define the unions as `oneOf` rather than `anyOf`, which again improves the error messages.

In any case, you aren't going to get static type checking of your Union types unless you define them explicitly, e.g. 

```python
UnionGetModels = XResponse | YResponse | ZResponse
```

So you need to decide if getting type hints while writing code is worth maintaining that. Once you put function calls or anything dynamic into the type hints then static analysis tools will not work. But, you can use `typing.TYPE_CHECKING` to define things statically for type checking / hints, and dynamically for pydantic as above. 

You might be find discussion https://github.com/pydantic/pydantic/discussions/8789 useful, although it is focussed more on statically defined model classes.
